<?xml version="1.0"?>
<launch>
  <!-- Mapping -->
  
    <arg name="rtabmap_args"            default="--delete_db_on_start"/>
    <arg name="approx_sync" 			default="false"/>
  	<arg name="left_image_topic"        default="visguide/zed_node/left/image_rect_color" />
  	<arg name="right_image_topic"       default="visguide/zed_node/right/image_rect" />      <!-- using grayscale image for efficiency -->
  	<arg name="left_camera_info_topic"  default="visguide/zed_node/left/camera_info" />
	<arg name="right_camera_info_topic" default="visguide/zed_node/right/camera_info" />
	<arg name="odom_args" default=""/>
	<arg name="launch_prefix" default=""/>
   
    <node  pkg="rtabmap_ros" type="stereo_odometry" name="stereo_odometry" output="screen" args="$(arg rtabmap_args) $(arg odom_args)" launch-prefix="$(arg launch_prefix)">
          <remap from="left/image_rect"        to="visguide/zed_node/left/image_rect_color"/>
          <remap from="right/image_rect"       to="visguide/zed_node/right/image_rect"/>
          <remap from="left/camera_info"       to="visguide/zed_node/left/camera_info"/>
          <remap from="right/camera_info"      to="visguide/zed_node/right/camera_info"/>
          <remap from="rgbd_image"             to="/visguide/zed_node/rgb_raw/image_raw_color"/>
          <remap from="odom"                   to="/odom"/>
          <remap from="imu"                    to=""/>
      
          <param name="frame_id"                    type="string" value="base_link"/>
          <param name="odom_frame_id"               type="string" value="odom"/>
          <param name="publish_tf"                  type="bool"   value="true"/>
          <param name="ground_truth_frame_id"       type="string" value=""/>
          <param name="ground_truth_base_frame_id"  type="string" value=""/>
          <param name="wait_for_transform_duration" type="double" value="0.2"/>
          <param name="wait_imu_to_init"            type="bool"   value="false"/>
          <param name="approx_sync"                 type="bool"   value="$(arg approx_sync)"/>
          <param name="config_path"                 type="string" value=""/>
          <param name="queue_size"                  type="int"    value="10"/>
          <param name="subscribe_rgbd"              type="bool"   value="false"/>
          <param name="guess_frame_id"              type="string" value=""/>
          <param name="guess_min_translation"       type="double" value="0"/>
          <param name="guess_min_rotation"          type="double" value="0"/>
</node>
  <group ns="rtabmap">
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap"  args="$(arg rtabmap_args)" output="screen">
          <param name="database_path"       type="string" value="rtabmap.db"/>
          <param name="frame_id"            type="string" value="base_link"/>
          <param name="subscribe_depth"     type="bool"   value="true"/>
          <param name="subscribe_scan"      type="bool"   value="false"/>
          <param name="map_negative_poses_ignored" type="bool" value="true"/>

          <!-- inputs -->
          <remap from="scan"            to="/scan"/>
          <remap from="rgb/image"       to="/visguide/zed_node/rgb_raw/image_raw_color"/>
          <remap from="depth/image"     to="/visguide/zed_node/depth/depth_registered"/>
          <remap from="rgb/camera_info" to="/visguide/zed_node/rgb_raw/camera_info"/>
          <remap from="odom"            to="/visguide/zed_node/odom"/>

          <!-- output -->
          <!-- <remap from="grid_map" to="/map"/> -->

          <!-- RTAB-Map's parameters: do "rosrun rtabmap rtabmap (double-dash)params" to see the list of available parameters. -->
          <param name="RGBD/ProximityBySpace"        type="string" value="true"/>   <!-- Local loop closure detection (using estimated position) with locations in WM -->
          <param name="RGBD/OptimizeFromGraphEnd"    type="string" value="false"/>  <!-- Set to false to generate map correction between /map and /odom -->
          <!-- <param name="Kp/MaxDepth"                  type="string" value="4.0"/> -->
          <!-- <param name="Reg/Strategy"                 type="string" value="0"/> -->      <!-- Loop closure transformation: 0=Visual, 1=ICP, 2=Visual+ICP -->
          <!-- <param name="Icp/CorrespondenceRatio"      type="string" value="0.3"/> -->
          <!-- <param name="Vis/MinInliers"               type="string" value="15"/>  -->     <!-- 3D visual words minimum inliers to accept loop closure -->
          <!-- <param name="Vis/InlierDistance"           type="string" value="0.1"/> -->    <!-- 3D visual words correspondence distance -->
          <param name="RGBD/AngularUpdate"           type="string" value="0.05"/>    <!-- Update map only if the robot is moving -->
          <param name="RGBD/LinearUpdate"            type="string" value="0.05"/>    <!-- Update map only if the robot is moving -->
          <!-- <param name="RGBD/ProximityPathMaxNeighbors" type="string" value="0"/> -->
          <param name="Mem/RehearsalSimilarity"      type="string" value="0.1"/>
          <!-- <param name="Reg/Force3DoF"                type="string" value="true"/> -->
          <param name="Rtabmap/DetectionRate"     type="string" value="1.0" />
          <param name="RGBD/OptimizeMaxError"        type="string" value="0"/>
          <param name="queue_size" type="int" value="30"/>
          <param name="Grid/FromDepth"               type="string" value="true"/>
          <param name="Grid/MaxObstacleHeight"    type="string" value="0.5" />
          <param name="Grid/NoiseFilteringRadius" type="string" value="0.1"/>
          <param name="Grid/NoiseFilteringMinNeighbors" type="string" value="10"/>
    </node>

  </group>
</launch>
